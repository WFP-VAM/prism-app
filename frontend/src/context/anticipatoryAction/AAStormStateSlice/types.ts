// // na/ny are not actually found in CSV, but defined not to cause confusion when calling the functions
import { DateItem } from 'config/types';
import { ParsedStormData } from './parsedStormDataTypes';
import { WindState } from './rawStormDataTypes';

export type ShortDate = string;
export type CycloneName = string;
export type TimeAndState = { ref_time: string; state: WindState };
export type AAStormWindStateReports = Record<
  ShortDate,
  Record<CycloneName, TimeAndState[]>
>;

export const AAPhase = ['Active', 'na'] as const;
export type AAPhaseType = (typeof AAPhase)[number];

export type AnticipatoryActionState = {
  data: ParsedStormData;
  windStateReports: AAStormWindStateReports;
  selectedStormName: string | undefined;
  // availableDates used to update layer available dates after csv processed
  availableDates?: DateItem[] | undefined;
  loading: boolean;
  error: string | null;
};
