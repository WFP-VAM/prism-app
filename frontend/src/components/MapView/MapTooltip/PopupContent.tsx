import {
  Typography,
  WithStyles,
  createStyles,
  withStyles,
} from '@material-ui/core';
import { ClassNameMap } from '@material-ui/styles';
import { PopupData, PopupMetaData } from 'context/tooltipStateSlice';
import { Position } from 'geojson';
import { useSafeTranslation } from 'i18n';
import { isEmpty, isEqual, sum } from 'lodash';
import React, { Fragment, memo } from 'react';
import { TFunction } from 'utils/data-utils';

const styles = () =>
  createStyles({
    phasePopulationTable: {
      tableLayout: 'fixed',
      borderCollapse: 'collapse',
      width: '100%',
      borderWidth: '1px;',
      borderColor: 'inherit',
      borderStyle: 'solid',
      border: '1px solid white',
    },
    phasePopulationTableRow: {
      border: '1px solid white',
    },
    text: {
      marginBottom: '4px',
    },
  });

// This function prepares phasePopulationTable for rendering and is specific
// to the data structure of the phase classification layer.
// Note - this is a bit hacky for now and will likely need to be revamped if we
// encounter other complex needs for tooltips.
const generatePhasePopulationTable = (
  popupData: PopupData,
  t: TFunction,
  classes: ClassNameMap,
): React.JSX.Element | null => {
  const phasePopulations: Record<string, number> = Object.entries(
    popupData,
  ).reduce((acc: any, cur: any) => {
    const [key, value] = cur;
    if (key.includes('Population in phase ')) {
      // extract number from string looking like "Phase classification 1"
      const phaseNumber = key.replace('Population in phase ', '');
      if (phaseNumber) {
        return { ...acc, [phaseNumber]: value.data };
      }
    }
    return acc;
  }, {});

  if (isEmpty(phasePopulations)) {
    return null;
  }

  // calculate total population
  // eslint-disable-next-line no-param-reassign, fp/no-mutation
  phasePopulations.Total =
    sum(Object.values(phasePopulations)) - phasePopulations['3 to 5'];

  const phasePopulationTable = (
    <div>
      <Typography display="inline" variant="h4" color="inherit">
        {t('Ref. period')}: {popupData['Reference period start']?.data} -{' '}
        {popupData['Reference period end']?.data}
      </Typography>
      <Typography variant="h4" color="inherit">
        {t('Population and percentage by phase classification')}
      </Typography>
      <table className={classes.phasePopulationTable}>
        <tr className={classes.phasePopulationTableRow}>
          {Object.keys(phasePopulations).map((phaseName: string) => (
            <th>{t(phaseName)}</th>
          ))}
        </tr>
        <tr className={classes.phasePopulationTableRow}>
          {Object.values(phasePopulations).map((populationInPhase: number) => (
            <th>{populationInPhase.toLocaleString()}</th>
          ))}
        </tr>
        <tr className={classes.phasePopulationTableRow}>
          {Object.values(phasePopulations).map((populationInPhase: number) => (
            <th>
              {Math.round((populationInPhase / phasePopulations.Total) * 100)}%
            </th>
          ))}
        </tr>
      </table>
    </div>
  );

  return phasePopulationTable;
};

interface PopupContentProps extends WithStyles<typeof styles> {
  popupData: PopupData & PopupMetaData;
  coordinates: Position | undefined;
}

const PopupContent = ({
  popupData,
  coordinates,
  classes,
}: PopupContentProps) => {
  const { t } = useSafeTranslation();

  const phasePopulationTable = generatePhasePopulationTable(
    popupData,
    t,
    classes,
  );
  // filter out popupData where key value contains "Population in phase "
  const popupDataWithoutPhasePopulations: PopupData = Object.entries(
    popupData,
  ).reduce((acc: any, cur: any) => {
    const [key, value] = cur;
    if (
      // keep "Population in phase 1" as a placeholder for the phase population table
      key === 'Population in phase 1' ||
      (!key.includes('Population in phase ') &&
        !key.includes('Reference period '))
    ) {
      return { ...acc, [key]: value };
    }
    return acc;
  }, {});

  return (
    <>
      {Object.entries(popupDataWithoutPhasePopulations)
        .filter(([, value]) => {
          // TODO - this seems like a hacky way to filter data and likely to break
          // Temporarily add logging to determine if this is actually filtering out any data
          // Note - introduced  by Harry in https://github.com/WFP-VAM/prism-app/pull/834/
          if (!isEqual(value.coordinates, coordinates)) {
            /* eslint-disable no-console */
            console.log(
              'Coordinates are not equal and some data should be omitted!',
            );
            console.log('Data Coordinates:', value.coordinates);
            console.log('Popup coordinates:', coordinates);
            /* eslint-enable no-console */
          }
          if (value.data === undefined) {
            return false;
          }
          // return isEqual(value.coordinates, coordinates);
          return true;
        })
        .map(([key, value]) => {
          return (
            <Fragment key={key}>
              {/* Allow users to show data without a key/title */}
              {!key.includes('do_not_display') &&
                key !== 'Population in phase 1' && (
                  <Typography
                    display="inline"
                    variant="h4"
                    color="inherit"
                    className={classes.text}
                  >
                    {`${t(key)}: `}
                  </Typography>
                )}
              {key !== 'Population in phase 1' && (
                <Typography
                  display="inline"
                  variant="h4"
                  color="inherit"
                  className={classes.text}
                >
                  {`${value.data}`}
                </Typography>
              )}
              {/* Phase classification data */}
              <Typography variant="h4" color="inherit">
                {value.adminLevel && `${t('Admin Level')}: ${value.adminLevel}`}
              </Typography>
              {key === 'Population in phase 1' && phasePopulationTable}
            </Fragment>
          );
        })}
    </>
  );
};

export default memo(withStyles(styles)(PopupContent));
