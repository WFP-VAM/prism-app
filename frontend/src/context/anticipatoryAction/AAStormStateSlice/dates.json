{
  "2024-12-09": {
    "04-20242025": [
      {
        "ref_time": "2024-12-09T06:00:00Z",
        "state": "monitoring",
        "path": "04-20242025/2024-12-09T06:00:00Z.json"
      },
      {
        "ref_time": "2024-12-09T12:00:00Z",
        "state": "monitoring",
        "path": "04-20242025/2024-12-09T12:00:00Z.json"
      },
      {
        "ref_time": "2024-12-09T18:00:00Z",
        "state": "monitoring",
        "path": "04-20242025/2024-12-09T18:00:00Z.json"
      }
    ]
  },
  "2024-12-10": {
    "chido": [
      {
        "ref_time": "2024-12-10T00:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-10T00:00:00Z.json"
      },
      {
        "ref_time": "2024-12-10T06:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-10T06:00:00Z.json"
      },
      {
        "ref_time": "2024-12-10T12:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-10T12:00:00Z.json"
      },
      {
        "ref_time": "2024-12-10T18:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-10T18:00:00Z.json"
      }
    ]
  },
  "2024-12-11": {
    "chido": [
      {
        "ref_time": "2024-12-11T00:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-11T00:00:00Z.json"
      },
      {
        "ref_time": "2024-12-11T06:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-11T06:00:00Z.json"
      },
      {
        "ref_time": "2024-12-11T12:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-11T12:00:00Z.json"
      },
      {
        "ref_time": "2024-12-11T18:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-11T18:00:00Z.json"
      }
    ]
  },
  "2024-12-12": {
    "chido": [
      {
        "ref_time": "2024-12-12T00:00:00Z",
        "state": "activated_64kt",
        "path": "chido/2024-12-12T00:00:00Z.json"
      },
      {
        "ref_time": "2024-12-12T06:00:00Z",
        "state": "activated_64kt",
        "path": "chido/2024-12-12T06:00:00Z.json"
      },
      {
        "ref_time": "2024-12-12T12:00:00Z",
        "state": "activated_64kt",
        "path": "chido/2024-12-12T12:00:00Z.json"
      },
      {
        "ref_time": "2024-12-12T18:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-12T18:00:00Z.json"
      }
    ]
  },
  "2024-12-13": {
    "chido": [
      {
        "ref_time": "2024-12-13T00:00:00Z",
        "state": "activated_64kt",
        "path": "chido/2024-12-13T00:00:00Z.json"
      },
      {
        "ref_time": "2024-12-13T06:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-13T06:00:00Z.json"
      },
      {
        "ref_time": "2024-12-13T12:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-13T12:00:00Z.json"
      },
      {
        "ref_time": "2024-12-13T18:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-13T18:00:00Z.json"
      }
    ]
  },
  "2024-12-14": {
    "chido": [
      {
        "ref_time": "2024-12-14T00:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-14T00:00:00Z.json"
      },
      {
        "ref_time": "2024-12-14T06:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-14T06:00:00Z.json"
      },
      {
        "ref_time": "2024-12-14T12:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-14T12:00:00Z.json"
      },
      {
        "ref_time": "2024-12-14T18:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-14T18:00:00Z.json"
      }
    ]
  },
  "2024-12-15": {
    "chido": [
      {
        "ref_time": "2024-12-15T00:00:00Z",
        "state": "ready",
        "path": "chido/2024-12-15T00:00:00Z.json"
      },
      {
        "ref_time": "2024-12-15T06:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-15T06:00:00Z.json"
      },
      {
        "ref_time": "2024-12-15T12:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-15T12:00:00Z.json"
      },
      {
        "ref_time": "2024-12-15T18:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-15T18:00:00Z.json"
      }
    ]
  },
  "2024-12-16": {
    "chido": [
      {
        "ref_time": "2024-12-16T00:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-16T00:00:00Z.json"
      },
      {
        "ref_time": "2024-12-16T06:00:00Z",
        "state": "monitoring",
        "path": "chido/2024-12-16T06:00:00Z.json"
      }
    ]
  },
  "2025-01-06": {
    "05-20242025": [
      {
        "ref_time": "2025-01-06T12:00:00Z",
        "state": "monitoring",
        "path": "05-20242025/2025-01-06T12:00:00Z.json"
      }
    ]
  },
  "2025-01-07": {
    "05-20242025": [
      {
        "ref_time": "2025-01-07T06:00:00Z",
        "state": "monitoring",
        "path": "05-20242025/2025-01-07T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-07T12:00:00Z",
        "state": "monitoring",
        "path": "05-20242025/2025-01-07T12:00:00Z.json"
      }
    ]
  },
  "2025-01-08": {
    "05-20242025": [
      {
        "ref_time": "2025-01-08T00:00:00Z",
        "state": "monitoring",
        "path": "05-20242025/2025-01-08T00:00:00Z.json"
      },
      {
        "ref_time": "2025-01-08T06:00:00Z",
        "state": "monitoring",
        "path": "05-20242025/2025-01-08T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-08T12:00:00Z",
        "state": "monitoring",
        "path": "05-20242025/2025-01-08T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-08T18:00:00Z",
        "state": "monitoring",
        "path": "05-20242025/2025-01-08T18:00:00Z.json"
      }
    ]
  },
  "2025-01-09": {
    "dikeledi": [
      {
        "ref_time": "2025-01-09T00:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-09T00:00:00Z.json"
      },
      {
        "ref_time": "2025-01-09T06:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-09T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-09T12:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-09T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-09T18:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-09T18:00:00Z.json"
      }
    ]
  },
  "2025-01-10": {
    "dikeledi": [
      {
        "ref_time": "2025-01-10T00:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-10T00:00:00Z.json"
      },
      {
        "ref_time": "2025-01-10T06:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-10T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-10T12:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-10T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-10T18:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-10T18:00:00Z.json"
      }
    ]
  },
  "2025-01-11": {
    "dikeledi": [
      {
        "ref_time": "2025-01-11T00:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-11T00:00:00Z.json"
      },
      {
        "ref_time": "2025-01-11T06:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-11T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-11T12:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-11T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-11T18:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-11T18:00:00Z.json"
      }
    ]
  },
  "2025-01-12": {
    "dikeledi": [
      {
        "ref_time": "2025-01-12T00:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-12T00:00:00Z.json"
      },
      {
        "ref_time": "2025-01-12T06:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-12T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-12T12:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-12T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-12T18:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-12T18:00:00Z.json"
      }
    ]
  },
  "2025-01-13": {
    "dikeledi": [
      {
        "ref_time": "2025-01-13T00:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-13T00:00:00Z.json"
      },
      {
        "ref_time": "2025-01-13T06:00:00Z",
        "state": "activated_64kt",
        "path": "dikeledi/2025-01-13T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-13T12:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-13T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-13T18:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-13T18:00:00Z.json"
      }
    ]
  },
  "2025-01-14": {
    "dikeledi": [
      {
        "ref_time": "2025-01-14T06:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-14T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-14T12:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-14T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-14T18:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-14T18:00:00Z.json"
      }
    ]
  },
  "2025-01-15": {
    "dikeledi": [
      {
        "ref_time": "2025-01-15T06:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-15T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-15T12:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-15T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-15T18:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-15T18:00:00Z.json"
      }
    ]
  },
  "2025-01-16": {
    "dikeledi": [
      {
        "ref_time": "2025-01-16T06:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-16T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-16T12:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-16T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-16T18:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-16T18:00:00Z.json"
      }
    ]
  },
  "2025-01-17": {
    "dikeledi": [
      {
        "ref_time": "2025-01-17T06:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-17T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-17T12:00:00Z",
        "state": "monitoring",
        "path": "dikeledi/2025-01-17T12:00:00Z.json"
      }
    ]
  },
  "2025-01-24": {
    "06-20242025": [
      {
        "ref_time": "2025-01-24T12:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-24T12:00:00Z.json"
      }
    ]
  },
  "2025-01-25": {
    "06-20242025": [
      {
        "ref_time": "2025-01-25T06:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-25T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-25T12:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-25T12:00:00Z.json"
      }
    ]
  },
  "2025-01-26": {
    "06-20242025": [
      {
        "ref_time": "2025-01-26T06:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-26T06:00:00Z.json"
      }
    ]
  },
  "2025-01-27": {
    "06-20242025": [
      {
        "ref_time": "2025-01-27T12:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-27T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-27T18:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-27T18:00:00Z.json"
      }
    ]
  },
  "2025-01-28": {
    "06-20242025": [
      {
        "ref_time": "2025-01-28T00:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-28T00:00:00Z.json"
      },
      {
        "ref_time": "2025-01-28T06:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-28T06:00:00Z.json"
      },
      {
        "ref_time": "2025-01-28T12:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-28T12:00:00Z.json"
      },
      {
        "ref_time": "2025-01-28T18:00:00Z",
        "state": "monitoring",
        "path": "06-20242025/2025-01-28T18:00:00Z.json"
      }
    ]
  },
  "2025-01-29": {
    "elvis": [
      {
        "ref_time": "2025-01-29T06:00:00Z",
        "state": "monitoring",
        "path": "elvis/2025-01-29T06:00:00Z.json"
      }
    ]
  }
}
