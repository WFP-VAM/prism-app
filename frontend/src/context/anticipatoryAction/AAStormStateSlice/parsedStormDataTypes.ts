import { Feature, Geometry, GeoJsonProperties } from 'geojson';
import {
  ForecastDetails,
  TimeSeries,
} from 'prism-common/dist/types/anticipatory-action-storm/reportResponse';

export enum AACategory {
  Severe = 'Severe',
  Moderate = 'Moderate',
  Risk = 'Risk',
}

export enum AACategoryKey {
  Severe = 'exposed_area_64kt',
  Moderate = 'exposed_area_48kt',
  Proba = 'proba_48kt_20_5d',
}

export enum AACategoryLandfall {
  Severe = 'severe tropical storm',
  Moderate = 'moderate tropical storm',
}

export interface AAData {
  districtNames: string[];
  polygon: any;
}

export type DistrictDataType = {
  [key in AACategory]?: AAData;
};

export interface LandfallInfo {
  district: string;
  time: string[];
  severity: AACategory[];
}

export const AACategoryDataToLandfallMap: {
  [key in AACategoryLandfall]: AACategory;
} = {
  [AACategoryLandfall.Severe]: AACategory.Severe,
  [AACategoryLandfall.Moderate]: AACategory.Moderate,
};

export const AACategoryKeyToCategoryMap: {
  [key in AACategoryKey]: AACategory;
} = {
  [AACategoryKey.Severe]: AACategory.Severe,
  [AACategoryKey.Moderate]: AACategory.Moderate,
  [AACategoryKey.Proba]: AACategory.Risk,
};

export interface MergedFeatures<P = GeoJsonProperties>
  extends Feature<Geometry | null, P> {
  geometry: Geometry | null;
}

/* parsed storm data type */
export type ParsedStormData = {
  readiness?: boolean;
  activeDistricts?: DistrictDataType;
  naDistricts?: DistrictDataType;
  landfall?: LandfallInfo;
  timeSeries?: TimeSeries;
  landfallDetected?: boolean;
  forecastDetails?: ForecastDetails;
  uncertaintyCone?: GeoJSON.Geometry;
  mergedGeoJSON?: {
    type: string;
    features: MergedFeatures[];
  };
};

export type ResultType = {
  data: ParsedStormData;
};
