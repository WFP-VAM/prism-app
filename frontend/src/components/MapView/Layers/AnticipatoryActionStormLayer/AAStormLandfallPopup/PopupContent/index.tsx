import { createStyles, makeStyles, Typography } from '@material-ui/core';

import { LandfallInfo } from 'context/anticipatoryAction/AAStormStateSlice/parsedStormDataTypes';
import {
  formatLandfallDate,
  formatLandfallEstimatedLeadtime,
  formatLandfallTimeRange,
  formatReportDate,
} from '../../utils';

function PopupContent({ landfallInfo, reportDate }: PopupContentProps) {
  const classes = useStyles();

  return (
    <div className={classes.itemsContainer}>
      <Typography
        variant="body1"
        className={`${classes.text} ${classes.title}`}
      >
        Report date: {formatReportDate(reportDate)}
      </Typography>
      <div className={classes.itemContainer}>
        <Typography variant="body1" className={classes.text}>
          landfall <span className={classes.block}>estimated time</span>
        </Typography>
        <Typography
          variant="body1"
          className={`${classes.text} ${classes.textAlignRight}`}
        >
          {formatLandfallDate(landfallInfo.time)}
          <span className={classes.block}>
            {formatLandfallTimeRange(landfallInfo.time)}
          </span>
        </Typography>
      </div>
      <div className={classes.itemContainer}>
        <Typography variant="body1" className={classes.text}>
          landfall estimated
          <span className={classes.block}>leadtime</span>
        </Typography>
        <Typography
          variant="body1"
          className={`${classes.text} ${classes.textAlignRight}`}
        >
          {formatLandfallEstimatedLeadtime(landfallInfo.time, reportDate)}
        </Typography>
      </div>
      <div className={classes.itemContainer}>
        <Typography variant="body1" className={classes.text}>
          District impacted
          <span className={classes.block}>by landfall</span>
        </Typography>
        <Typography
          variant="body1"
          className={`${classes.text} ${classes.textAlignRight}`}
        >
          {landfallInfo.district}
        </Typography>
      </div>
    </div>
  );
}

const useStyles = makeStyles(() =>
  createStyles({
    block: {
      display: 'block',
    },
    itemsContainer: {
      display: 'flex',
      flexDirection: 'column',
      gap: '2px',
    },
    itemContainer: {
      display: 'flex',
      justifyContent: 'space-between',
      padding: '8px',
      background: '#FFFFFF',
    },
    text: {
      fontSize: '14px',
      lineHeight: '18px',
      fontWeight: 400,
    },
    title: {
      fontWeight: 700,
      padding: '2px 0px 2px 8px',
    },
    textAlignRight: {
      textAlign: 'right',
    },
  }),
);

interface PopupContentProps {
  landfallInfo: LandfallInfo;
  reportDate: string;
}

export default PopupContent;
