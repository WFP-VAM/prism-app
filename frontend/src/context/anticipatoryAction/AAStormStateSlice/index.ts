import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { DateItem } from 'config/types';
import type { CreateAsyncThunkTypes, RootState } from '../../store';
import { AAStormWindStateReports, AnticipatoryActionState } from './types';
import { parseAndTransformAA } from './utils';
import { ParsedStormData } from './parsedStormDataTypes';
import { StormDataResponseBody } from './rawStormDataTypes';

const initialState: AnticipatoryActionState = {
  data: {},
  windStateReports: {},
  availableDates: undefined,
  selectedStormName: undefined,
  loading: false,
  error: null,
};

export const loadAllAAStormData = createAsyncThunk<
  boolean,
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionStormState/loadAllData', async (_, { dispatch }) => {
  dispatch(loadWindStateReports());
  return true;
});

export const loadLatestStormReport = createAsyncThunk<
  {
    data: ParsedStormData;
  },
  undefined,
  CreateAsyncThunkTypes
>(
  'anticipatoryActionStormState/loadLatestStormReport',
  async (_, { rejectWithValue }) => {
    try {
      const response = await fetch(
        'https://data.earthobservation.vam.wfp.org/public-share/aa/ts/outputs/latest.json',
      );
      const stormData = await response.json();
      const data = parseAndTransformAA(stormData as StormDataResponseBody);
      return data;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const loadStormReport = createAsyncThunk<
  {
    data: ParsedStormData;
  },
  { stormName: string; date: string },
  CreateAsyncThunkTypes
>(
  'anticipatoryActionStormState/loadStormReport',
  async ({ stormName, date }, { rejectWithValue }) => {
    if (!stormName || !date) {
      // eslint-disable-next-line no-console
      console.warn('Storm name and date are required');
      return rejectWithValue(new Error('Storm name and date are required'));
    }
    try {
      const response = await fetch(
        `https://data.earthobservation.vam.wfp.org/public-share/aa/ts/outputs/${stormName}/${date}.json?v2`,
      );
      const stormData = await response.json();
      const data = parseAndTransformAA(stormData as StormDataResponseBody);
      return data;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

// Update the DateItem type import or extend it locally
export type ExtendedDateItem = DateItem & {
  stormNames: string[];
};

export const loadWindStateReports = createAsyncThunk<
  { reports: AAStormWindStateReports; availableDates: ExtendedDateItem[] },
  undefined,
  CreateAsyncThunkTypes & {
    rejectValue: unknown;
  }
>(
  'anticipatoryActionStormState/loadWindStateReports',
  async (_, { rejectWithValue }) => {
    try {
      const response = await fetch(
        'https://data.earthobservation.vam.wfp.org/public-share/aa/ts/outputs/dates.json',
      );
      const responseData = await response.json();
      const availableDates = Object.keys(responseData).map(dateStr => {
        // Parse the date string as UTC
        // eslint-disable-next-line prefer-template
        const utcDate = new Date(dateStr + 'T12:00:00.000Z');
        const timestamp = utcDate.getTime();
        return {
          displayDate: timestamp,
          queryDate: timestamp,
          startDate: timestamp,
          endDate: timestamp,
          stormNames: Object.keys(responseData[dateStr]),
        };
      });
      return {
        reports: responseData,
        availableDates,
      };
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const anticipatoryActionStormStateSlice = createSlice({
  name: 'anticipatoryActionStormState',
  initialState,
  reducers: {
    setSelectedStormName: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['selectedStormName']>,
    ) => ({
      ...state,
      selectedStormName: payload,
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadStormReport.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      data: payload.data,
    }));

    builder.addCase(loadStormReport.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadStormReport.pending, ({ error: _error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
    builder.addCase(loadWindStateReports.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      windStateReports: payload.reports,
      availableDates: payload.availableDates,
    }));

    builder.addCase(loadWindStateReports.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(
      loadWindStateReports.pending,
      ({ error: _error, ...state }) => ({
        ...state,
        error: null,
        loading: true,
      }),
    );
  },
});

// export selectors
export const AADataSelector = (state: RootState) =>
  state.anticipatoryActionStormState.data;

export const AALoadingSelector = (state: RootState) =>
  state.anticipatoryActionStormState.loading;

export const AAAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionStormState.availableDates;

export const AAWindStateReports = (state: RootState) =>
  state.anticipatoryActionStormState.windStateReports;

export const { setSelectedStormName } =
  anticipatoryActionStormStateSlice.actions;

export const AASelectedStormNameSelector = (state: RootState) =>
  state.anticipatoryActionStormState.selectedStormName;

export default anticipatoryActionStormStateSlice.reducer;
