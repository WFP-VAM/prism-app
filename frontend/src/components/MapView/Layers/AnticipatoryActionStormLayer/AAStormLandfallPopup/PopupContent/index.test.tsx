import { render } from '@testing-library/react';

import { AACategory } from 'context/anticipatoryAction/AAStormStateSlice/parsedStormDataTypes';
import PopupContent from '.';

describe('AAStormLandfallPopup component', () => {
  it('renders as expected', () => {
    const reportDate = '2024-03-01 18:00:00';
    const landfallInfo = {
      time: ['2024-03-12 01:00:00', '2024-03-12 06:00:00'],
      district: 'Inhassoro',
      severity: [AACategory.Severe, AACategory.Moderate],
    };

    const { container } = render(
      <PopupContent reportDate={reportDate} landfallInfo={landfallInfo} />,
    );
    expect(container).toMatchSnapshot();
  });
});
