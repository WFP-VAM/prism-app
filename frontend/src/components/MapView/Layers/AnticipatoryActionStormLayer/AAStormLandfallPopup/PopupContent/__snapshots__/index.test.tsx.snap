// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AAStormLandfallPopup component renders as expected 1`] = `
<div>
  <div
    class="makeStyles-itemsContainer-2"
  >
    <mock-typography
      classname="makeStyles-text-4 makeStyles-title-5"
      variant="body1"
    >
      Report date: 
      2024-03-01 8pm (GMT+2)
    </mock-typography>
    <div
      class="makeStyles-itemContainer-3"
    >
      <mock-typography
        classname="makeStyles-text-4"
        variant="body1"
      >
        landfall 
        <span
          class="makeStyles-block-1"
        >
          estimated time
        </span>
      </mock-typography>
      <mock-typography
        classname="makeStyles-text-4 makeStyles-textAlignRight-6"
        variant="body1"
      >
        2024-03-12 03:00 GMT+2
        <span
          class="makeStyles-block-1"
        >
          += 5 hours
        </span>
      </mock-typography>
    </div>
    <div
      class="makeStyles-itemContainer-3"
    >
      <mock-typography
        classname="makeStyles-text-4"
        variant="body1"
      >
        landfall estimated
        <span
          class="makeStyles-block-1"
        >
          leadtime
        </span>
      </mock-typography>
      <mock-typography
        classname="makeStyles-text-4 makeStyles-textAlignRight-6"
        variant="body1"
      >
        247 - 252 hrs
      </mock-typography>
    </div>
    <div
      class="makeStyles-itemContainer-3"
    >
      <mock-typography
        classname="makeStyles-text-4"
        variant="body1"
      >
        District impacted
        <span
          class="makeStyles-block-1"
        >
          by landfall
        </span>
      </mock-typography>
      <mock-typography
        classname="makeStyles-text-4 makeStyles-textAlignRight-6"
        variant="body1"
      >
        Inhassoro
      </mock-typography>
    </div>
  </div>
</div>
`;
